### PR TITLE
Optional Arguments for JavaScript callbacks

### PR DESCRIPTION
**Pull request by Schneegans**

### CODE REVIEW

File: src/cs-gui/WebView.hpp [hunk 1]
```diff
@@ -168,14 +165,38 @@ class CS_GUI_EXPORT WebView {
...
+        "(and std::optionals thereof)!");
   }
 
-  /// This wraps the given callback in a lambda which will stored in an internal map. This lambda
-  /// receives its arguments as a std::vector<std::any>, each item in this vector will be casted to
-  /// the required paramater types of the given callback.
+  /// The UnderlyingValue struct is used to access the actual value in a std::optional<JSType>.
+  /// There are two variants of the struct as we may want to have the actual value (a bool, double
+  /// or std::string) contained in the std::optional<JSType>, or an std::optional thereof (either a
+  /// std::optional<bool>, std::optional<double>, or std::optional<std::string>).
```
Comment (TheJhonny007) [inline src/cs-gui/WebView.hpp]:
```suggestion
  /// The UnderlyingValue struct is used to access the actual value in an std::optional<JSType>.
  /// There are two variants of the struct as we may want to have the actual value (a bool, double
  /// or std::string) contained in the std::optional<JSType>, or an std::optional thereof (either an
  /// std::optional<bool>, std::optional<double>, or std::optional<std::string>).
```

File: src/cs-gui/WebView.hpp [hunk 2]
```diff
@@ -168,14 +165,38 @@ class CS_GUI_EXPORT WebView {
...
+    static inline constexpr std::optional<T> get(std::optional<JSType>&& value) {
+      if (value.has_value()) {
+        return std::get<T>(std::move(value.value()));
+      }
+      return std::nullopt;
+    }
+  };
+
+  /// This wraps the given callback in a lambda which will be stored in an internal map. This lambda
+  /// receives its arguments as a std::vector<std::optional<JSType>>, each item in this vector will
```
Comment (TheJhonny007) [inline src/cs-gui/WebView.hpp]:
```suggestion
  /// receives its arguments as an std::vector<std::optional<JSType>>, each item in this vector will
```

File: src/cs-gui/internal/WebViewClient.cpp [hunk 3]
```diff
@@ -16,13 +16,13 @@ namespace cs::gui::detail {
...
-  if (js_callbacks_.size() > 0) {
+  if (mJSCallbacks.size() > 0) {
     spdlog::warn(
         "While destructing a WebViewClient there were still JavaScript callbacks registered:");
 
-    for (auto&& i : js_callbacks_) {
+    for (auto&& i : mJSCallbacks) {
       spdlog::warn(" - {}", i.first);
-      i.second = [](std::vector<std::any> const&) {};
+      i.second = [](std::vector<std::optional<JSType>> const&) {};
```
Comment (TheJhonny007) [inline src/cs-gui/internal/WebViewClient.cpp]:
```suggestion
      i.second = [](auto) {};
```

File: src/cosmoscout/Application.cpp [hunk 4]
```diff
@@ -871,26 +871,30 @@ void Application::registerGuiCallbacks() {
...
+      "absolute difference to the current simulation time is lower than the given threshold "
+      "(optionalDouble2, default is 172800s which is 48h), there will be a transition of the given "
+      "duration (optionalDouble, default is 0s).",
+      std::function(
+          [this](double tTime, std::optional<double> duration, std::optional<double> threshold) {
+            mTimeControl->setTime(tTime, duration.value_or(0.0), threshold.value_or(48 * 60 * 60));
+          }));
 
   mGuiManager->getGui()->registerCallback("time.reset",
       "Resets the simulation time to the default value.",
```
[the remark this answers is not in the document]
Comment (Schneegans) [reply src/cosmoscout/Application.cpp]: Done.

File: src/cs-core/TimeControl.cpp [hunk 5]
```diff
@@ -33,6 +33,16 @@ TimeControl::~TimeControl() {
...
 void TimeControl::update() {
+  // Initialize our members. This has to be done here as SPICE is not yet loaded at construction
+  // time.
+  if (mStartDate == "") {
+    mStartDate = mSettings->mStartDate;
+    mMaxDate =
+        cs::utils::convert::toSpiceTime(boost::posix_time::time_from_string(mSettings->mMaxDate));
+    mMinDate =
+        cs::utils::convert::toSpiceTime(boost::posix_time::time_from_string(mSettings->mMinDate));
+  }
```
[the remark this answers is not in the document]
Comment (Schneegans) [reply src/cs-core/TimeControl.cpp]: Done.

File: src/cs-gui/WebView.hpp (hunk 1 above)
Comment (Schneegans) [reply src/cs-gui/WebView.hpp]: I used "a std::..." consistently.

File: src/cs-gui/WebView.hpp (hunk 2 above)
Comment (Schneegans) [reply src/cs-gui/WebView.hpp]: I used "a std::..." consistently.

File: src/cs-gui/internal/WebViewClient.cpp (hunk 3 above)
Comment (Schneegans) [reply src/cs-gui/internal/WebViewClient.cpp]: Done.

File: src/cs-gui/WebView.hpp [hunk 6]
```diff
@@ -128,8 +128,8 @@ class CS_GUI_EXPORT WebView {
   virtual void goForward() const;
 
   /// These could be executed when the according hot keys are pressed. The system's clipboard will
-  /// be used. So the user can actually copy-paste from one WebView to another or even from an to
-  /// third-party applications.
+  /// be used. So the user can actually copy-paste from one WebView to another or even from a
+  /// WebView to a third-party applications.
```
Comment (TheJhonny007) [inline src/cs-gui/WebView.hpp]:
```suggestion
  /// WebView to third-party applications.
```

File: src/cs-gui/WebView.hpp [hunk 7]
```diff
@@ -194,7 +194,7 @@ class CS_GUI_EXPORT WebView {
     }
   };
 
-  /// This wraps the given callback in a lambda which will be stored in an internal map. This lambda
+  /// This wraps the given callback in a lambda which will be stored in a internal map. This lambda
```
Comment (TheJhonny007) [inline src/cs-gui/WebView.hpp]:
```suggestion
  /// This wraps the given callback in a lambda which will be stored in an internal map. This lambda
```

**Pull request merged**